### ISSUE TITLE
Unknown CA Authority for github.com

### ISSUE THREAD

**Issue by pimlottc-gov** [github/hub#1804]
On my MacBook Pro, running High Sierra, using `hub` installed via homebrew:

```
$ GITHUB_TOKEN=[REMOVED] hub pr list
Error fetching pull requests: Get https://api.github.com/repos/usdhs/asylum-assessment-generator/pulls?per_page=100&: x509: certificate signed by unknown authority
```

Note this is different than #852 as this project using ordinary github.com, which uses a normal TLS certificate signed by a well-known root CA.

**Comment by mislav** [comment]
Thank you for reporting! Does this happen every time? What is your hub version?

**Comment by pimlottc-gov** [comment]
Yes, it happens every time.  I just set up hub on a new computer.

```
$ hub --version
git version 2.17.1
hub version 2.5.0
```

**Comment by mislav** [comment]
Can you save this as a file anywhere, e.g. `pbpaste > http.rb`

```go
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
)

func main() {
	api := http.Client{}
	req, _ := http.NewRequest("GET", "https://api.github.com/", nil)
	res, err := api.Do(req)
	if err != nil {
		panic(err)
	}
	fmt.Fprintf(os.Stdout, "HTTP %d\n", res.StatusCode)
	io.Copy(os.Stdout, res.Body)
	fmt.Fprintf(os.Stdout, "\n")
}
```

And run it with Go (`brew install go`): `go run ./http.go`

Does that work?

**Comment by pimlottc-gov** [comment]
Running that gives me the same error:

```
$ go run ./http.go 
panic: Get https://api.github.com/: x509: certificate signed by unknown authority

goroutine 1 [running]:
main.main()
	/Users/pimlottc/code/go/http.go:15 +0x1e4
exit status 2
```

**Comment by pimlottc-gov** [comment]
Actually, I get the same error on many others sites (if I change the code in http.go):

* www.github.com
* www.google.com
* www.facebook.com
* www.apple.com
* www.reddit.com

However, these work:
* www.amazon.com

I can open all these curl/wget/chrome with no errors.

**Comment by mislav** [comment]
Sounds like something is wrong with the CA bundle on your machine. Could anything that you've installed/configured affected your system CA bundle?

Since a simple Go program reproduces this problem, I would wager that this is not something we can solve in hub.

**Comment by pimlottc-gov** [comment]
Possibly; I have made some changes.  But all these root CAs are in my Keychain, and other tools like Firefox, Chrome, Safari, curl, wget, python are all able to hit them without TLS errors.  From my brief googling, it looks like Go should be able to use the macOS Keychain?

This does seem to be outside the scope of hub, though :P

**Comment by pimlottc-gov** [comment]
I think this is the issue: https://github.com/golang/go/issues/24652 .  Many of my System Root CAs have custom trust settings as "Always Trust" which seems to be causing Go not to trust them.

**Comment by egyptiankarim** [comment]
I can confirm that running the Homebrew-installed `hub` (2.5.0) on High Sierra will work against Enterprise instances using my enterprise CA when the "System" Keychain is set to explicitly "Always Trust" that CA. That is, I am now able to run `hub` commands pointing at an Enterprise repository with no complaints about the certificates I'm using.

**Disclaimer**: In my case, it's an enterprise CA that I'm explicitly trusting, so I consider this slightly less risky than explicitly trusting a self-signed certificate, but everyone can/should do that calculation for themselves.

**Comment by espoelstra** [comment]
See also the recent comment on #852 where in the above golang ticket they mention just toggling trust back and forth can fix things.

**Comment by john-kurkowski** [comment]
> See also the recent comment on #852 where in the above golang ticket they mention just toggling trust back and forth can fix things.

[direct link](https://github.com/github/hub/issues/852#issuecomment-440459602)

**Comment by JoeGermuska** [comment]
I was also having this issue. 

I found the instructions (for people with issues against `api.github.com` as opposed to enterprise hostnames) unclear, because I didn't know which certs to toggle trust settings.

You can find it looking at the cert details if you load `https://api.github.com` in a browser, but in case it helps the next person, the cert is named **DigiCert High Assurance EV Root CA** -- in my Keychain Access app it's actually in two keychains:

![image](https://user-images.githubusercontent.com/46313/51619707-842f6f00-1ef6-11e9-94de-0e3539ebf8de.png)

Double-clicking on the name in that list opens this:
![image](https://user-images.githubusercontent.com/46313/51619596-534f3a00-1ef6-11e9-9545-b1a22bed2feb.png)

Expand the "Trust" settings by clicking the arrow, and set the menu to "never trust" and then close the window. You'll have to authenticate.  Open it again and switch it back.  I just did it on both keychains (System and System Roots) rather than experiment to see if it mattered which one to use.

Hope this helps the next person.

I don't know if it's related, but after fixing this, I encountered #1067 -- which I solved, as discussed there, by simply deleting `~/.config/hub` and reauthenticating.  (Note that when the prompt asks for your "github username" it actually wants the email address, not the display user name...)

**Comment by pimlottc-gov** [comment]
Unfortunately, I'm on a somewhat locked-down work computer and I'm not able to change trust settings in Keychain Access.

**Comment by espoelstra** [comment]
@pimlottc-gov does that include the keychain settings for your own user (login) rather than the System keychain?

**Comment by jkroes** [comment]
This happens on a work computer where our root certificate is self-signed. Git for Windows works with hub, but WSL fails with this error.

**Comment by DrNeuroStranGer13** [comment]
Try sending it  @users.noreply.github.com.

**Comment by DrNeuroStranGer13** [comment]
Peece i am outtie

**Comment by DrNeuroStranGer13** [comment]
Try This after your user name
(Username)@users.noreply.github.com.

**Comment by DrNeuroStranGer13** [comment]
Should work College Hackers
